// @grafana/toolkit/src/config/jest.plugin.config.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var path = require("path");
var fs_1 = tslib_1.__importDefault(require("fs"));
var whitelistedJestConfigOverrides = ['snapshotSerializers', 'moduleNameMapper'];
exports.jestConfig = function () {
    var jestConfigOverrides = require(path.resolve(process.cwd(), 'package.json')).jest;
    var blacklistedOverrides = jestConfigOverrides
        ? Object.keys(jestConfigOverrides).filter(function (override) { return whitelistedJestConfigOverrides.indexOf(override) === -1; })
        : [];
    if (blacklistedOverrides.length > 0) {
        console.error("\ngrafana-toolkit doesn't support following Jest options: ", blacklistedOverrides);
        console.log('Supported Jest options are: ', JSON.stringify(whitelistedJestConfigOverrides));
        throw new Error('Provided Jest config is not supported');
    }
    var shimsFilePath = path.resolve(process.cwd(), 'config/jest-shim.ts');
    var setupFilePath = path.resolve(process.cwd(), 'config/jest-setup.ts');
    var setupFile = fs_1.default.existsSync(setupFilePath) ? setupFilePath : undefined;
    var shimsFile = fs_1.default.existsSync(shimsFilePath) ? shimsFilePath : undefined;
    var setupFiles = [setupFile, shimsFile].filter(function (f) { return f; });
    var defaultJestConfig = {
        preset: 'ts-jest',
        verbose: false,
        moduleDirectories: ['node_modules', 'src'],
        moduleFileExtensions: ['ts', 'tsx', 'js', 'jsx', 'json'],
        setupFiles: setupFiles,
        globals: { 'ts-jest': { isolatedModules: true } },
        coverageReporters: ['json-summary', 'text', 'lcov'],
        collectCoverageFrom: ['src/**/*.{ts,tsx}', '!**/node_modules/**', '!**/vendor/**'],
        testMatch: [
            '<rootDir>/src/**/__tests__/**/*.{js,jsx,ts,tsx}',
            '<rootDir>/src/**/*.{spec,test,jest}.{js,jsx,ts,tsx}',
        ],
        transformIgnorePatterns: [
            '[/\\\\\\\\]node_modules[/\\\\\\\\].+\\\\.(js|jsx|ts|tsx)$',
            '^.+\\\\.module\\\\.(css|sass|scss)$',
        ],
    };
    return tslib_1.__assign({}, defaultJestConfig, jestConfigOverrides);
};
