// @grafana/toolkit/src/cli/tasks/changelog.js
"use strict";
var _this = this;
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
// @ts-ignore
var _ = tslib_1.__importStar(require("lodash"));
var task_1 = require("./task");
var githubClient_1 = tslib_1.__importDefault(require("../utils/githubClient"));
var changelogTaskRunner = function (_a) {
    var milestone = _a.milestone;
    return tslib_1.__awaiter(_this, void 0, void 0, function () {
        var githubClient, client, res, issues, bugs, notBugs, markdown, notBugs_1, notBugs_1_1, item, bugs_1, bugs_1_1, item;
        var e_1, _b, e_2, _c;
        return tslib_1.__generator(this, function (_d) {
            switch (_d.label) {
                case 0:
                    githubClient = new githubClient_1.default();
                    client = githubClient.client;
                    if (!/^\d+$/.test(milestone)) {
                        console.log('Use milestone number not title, find number in milestone url');
                        return [2 /*return*/];
                    }
                    return [4 /*yield*/, client.get('/issues', {
                            params: {
                                state: 'closed',
                                per_page: 100,
                                labels: 'add to changelog',
                                milestone: milestone,
                            },
                        })];
                case 1:
                    res = _d.sent();
                    issues = res.data;
                    bugs = _.sortBy(issues.filter(function (item) {
                        if (item.title.match(/fix|fixes/i)) {
                            return true;
                        }
                        if (item.labels.find(function (label) { return label.name === 'type/bug'; })) {
                            return true;
                        }
                        return false;
                    }), 'title');
                    notBugs = _.sortBy(issues.filter(function (item) { return !bugs.find(function (bug) { return bug === item; }); }), 'title');
                    markdown = '';
                    if (notBugs.length > 0) {
                        markdown = '### Features / Enhancements\n';
                    }
                    try {
                        for (notBugs_1 = tslib_1.__values(notBugs), notBugs_1_1 = notBugs_1.next(); !notBugs_1_1.done; notBugs_1_1 = notBugs_1.next()) {
                            item = notBugs_1_1.value;
                            markdown += getMarkdownLineForIssue(item);
                        }
                    }
                    catch (e_1_1) { e_1 = { error: e_1_1 }; }
                    finally {
                        try {
                            if (notBugs_1_1 && !notBugs_1_1.done && (_b = notBugs_1.return)) _b.call(notBugs_1);
                        }
                        finally { if (e_1) throw e_1.error; }
                    }
                    if (bugs.length > 0) {
                        markdown += '\n### Bug Fixes\n';
                    }
                    try {
                        for (bugs_1 = tslib_1.__values(bugs), bugs_1_1 = bugs_1.next(); !bugs_1_1.done; bugs_1_1 = bugs_1.next()) {
                            item = bugs_1_1.value;
                            markdown += getMarkdownLineForIssue(item);
                        }
                    }
                    catch (e_2_1) { e_2 = { error: e_2_1 }; }
                    finally {
                        try {
                            if (bugs_1_1 && !bugs_1_1.done && (_c = bugs_1.return)) _c.call(bugs_1);
                        }
                        finally { if (e_2) throw e_2.error; }
                    }
                    console.log(markdown);
                    return [2 /*return*/];
            }
        });
    });
};
function getMarkdownLineForIssue(item) {
    var githubGrafanaUrl = 'https://github.com/grafana/grafana';
    var markdown = '';
    var title = item.title.replace(/^([^:]*)/, function (_match, g1) {
        return "**" + g1 + "**";
    });
    markdown += '* ' + title + '.';
    markdown += " [#" + item.number + "](" + githubGrafanaUrl + "/pull/" + item.number + ")";
    markdown += ", [@" + item.user.login + "](" + item.user.html_url + ")";
    markdown += '\n';
    return markdown;
}
exports.changelogTask = new task_1.Task('Changelog generator task', changelogTaskRunner);
